lib/ProductOpener/APITaxonomySuggestions.pm

```diff
@@ -0,0 +1,556 @@
+# This file is part of Product Opener.
+#
+# Product Opener
+# Copyright (C) 2011-2023 Association Open Food Facts
+# Contact: contact@openfoodfacts.org
+# Address: 21 rue des Iles, 94100 Saint-Maur des Fossés, France
+#
+# Product Opener is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Affero General Public License as
+# published by the Free Software Foundation, either version 3 of the
+# License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+=head1 NAME
+
+ProductOpener::APITaxonomySuggestions - suggests taxonomy entries for dropdown, autocomplete etc.
+
+=head1 DESCRIPTION
+
+=cut
+
+package ProductOpener::APITaxonomySuggestions;
+
+use ProductOpener::PerlStandards;
+use Exporter qw< import >;
+
+use Log::Any qw($log);
+
+BEGIN {
+	use vars qw(@ISA @EXPORT_OK %EXPORT_TAGS);
+	@EXPORT_OK = qw(
+		&taxonomy_suggestions_api
+		&get_taxonomy_suggestions
+	);    # symbols to export on request
+	%EXPORT_TAGS = (all => [@EXPORT_OK]);
+}
+
+use vars @EXPORT_OK;
+
+use ProductOpener::Config qw/:all/;
+use ProductOpener::Store qw/:all/;
+use ProductOpener::Display qw/:all/;
+use ProductOpener::Lang qw/:all/;
+use ProductOpener::Tags qw/:all/;
+use ProductOpener::API qw/:all/;
+use ProductOpener::PackagerCodes qw/:all/;
+
+use List::Util qw/min/;
+use Data::DeepAccess qw(deep_exists deep_get);
+use Encode;
+
+=head2 taxonomy_suggestions_api ( $request_ref )
+
+Process API V3 taxonomy suggestions requests.
+
+=head3 Parameters
+
+=head4 $request_ref (input)
+
+Reference to the request object.
+
+=cut
+
+sub taxonomy_suggestions_api ($request_ref) {
```

Comment:
> I can't stand thinking would'nt it be better to tackle this problem using a specific mongodb collection ? (or elasticsearch ?).

It would be interesting to brainstorm. For mongodb, I'm not sure what we would put in the collection (1 doc = 1 taxonomy entry?) and how it would be queried. For ElasticSearch, I don't have experience with it so I don't know how well it would work for this problem and how to structure it.

That said, the problem is not simple: we want suggestions sorted by popularity for specific product characteristics (country, multiple categories), sometimes for sub-product components (packaging shapes and materials). We want to match synonyms of the taxonomy entries, in multiple languages. And we want almost instant responses (so we can't aggregate product data for each keystroke).